components/gitpod-db/src/typeorm/workspace-db-impl.ts

```diff
@@ -1040,50 +1040,87 @@ export class TypeORMWorkspaceDBImpl extends TransactionalDBImpl<WorkspaceDB> imp
     /**
      * Finds prebuilt workspaces by organization with optional filtering and pagination.
      * @param organizationId The ID of the organization.
-     * @param offset Offset for pagination.
-     * @param limit Limit for pagination.
+     * @param pagination Pagination per page size and result offset.
      * @param filter Filters for the search.
+     * @param sort Sort field and direction
      * @returns A promise that resolves to an array of PrebuiltWorkspace objects.
      */
     async findPrebuiltWorkspacesByOrganization(
         organizationId: string,
-        offset = 0,
-        limit = 25,
-        filter?: {
+        pagination: {
+            offset: number;
+            limit: number;
+        },
+        filter: {
             configuration?: {
                 id: string;
                 branch?: string;
             };
-            status: PrebuiltWorkspaceState;
+            state?: "succeeded" | "failed" | "unfinished";
             searchTerm?: string;
         },
+        sort: {
+            field: string;
+            order: "ASC" | "DESC";
+        },
     ): Promise<PrebuiltWorkspace[]> {
         const repo = await this.getPrebuiltWorkspaceRepo();
         const query = repo
             .createQueryBuilder("pws")
-            .orderBy("pws.creationTime", "DESC")
             .innerJoinAndMapOne(
                 "pws.workspace",
                 DBWorkspace,
                 "ws",
                 "pws.buildWorkspaceId = ws.id AND ws.organizationId = :organizationId",
                 { organizationId },
             )
-            .skip(offset)
-            .take(limit);
-
-        if (filter?.status) {
-            query.andWhere("pws.state = :state", { state: filter.status });
+            .skip(pagination.offset)
+            .take(pagination.limit)
+            .orderBy("pws.creationTime", sort.order); // todo: take sort field into account
+
+        if (filter.state) {
+            const { state } = filter;
+            // translating API state to DB state
+            switch (state) {
+                case "failed":
+                    query.andWhere(
+                        new Brackets((qb) => {
+                            const failedStates = ["failed", "aborted", "timeout"];
+                            qb.andWhere("pws.state IN (:...failedStates)", { failedStates }).orWhere(
```

Comment:
`:...` What does that operator do? Why not `(failedStates)? :thinking:

Reply:
Not sure if that would work, I'm following the example they have listed in their docs:

https://typeorm.io/select-query-builder#adding-where-expression

<img width="906" alt="image" src="https://github.com/gitpod-io/gitpod/assets/29888641/42937e14-41ba-4028-af87-6c8376592336">

Reply:
I assume it's to properly serialize it, but not actually sure